firedrake/assemble.py

```diff
@@ -277,12 +383,68 @@ def get_matrix(expr, mat_type, sub_mat_type, *, bcs=None,
         raise ValueError("Monolithic matrix assembly not supported for systems "
                          "with R-space blocks")
 
-    tensor = matrix.Matrix(expr, bcs, mat_type, sparsity, ScalarType,
-                           options_prefix=options_prefix)
-    return tensor, (), lambda: tensor
+    return matrix.Matrix(expr, bcs, opts.mat_type, sparsity, ScalarType,
+                         options_prefix=opts.options_prefix)
+
+
+def _assemble_expr(expr, tensor, bcs, opts, assembly_rank):
+    """Assemble an expression.
+
+    :arg expr: The expression to be assembled.
+    :arg tensor: The tensor to write to.
+    :arg bcs: Iterable of boundary conditions.
+    :arg opts: :class:`_AssemblyOpts` containing the assembly options.
+    :arg assembly_rank: The appropriate :class:`_AssemblyRank`.
+    """
+    parloop_init_args = (expr, tensor, bcs, opts.diagonal, opts.fc_params, assembly_rank)
+    cached_init_args, cached_parloops = expr._cache.get("parloops", (None, None))
+    parloops = cached_parloops if cached_init_args == parloop_init_args else None
+
+    if not parloops:
+        parloops = _make_parloops(*parloop_init_args)
+        expr._cache["parloops"] = (parloop_init_args, parloops)
+
+    for parloop in parloops:
+        parloop.compute()
+
+    dir_bcs = tuple(bc for bc in bcs if isinstance(bc, DirichletBC))
+    _apply_dirichlet_bcs(tensor, dir_bcs, opts, assembly_rank)
 
+    eq_bcs = tuple(bc for bc in bcs if isinstance(bc, EquationBCSplit))
+    if eq_bcs and opts.diagonal:
+        raise NotImplementedError("Diagonal assembly and EquationBC not supported")
+    for bc in eq_bcs:
+        if assembly_rank == _AssemblyRank.VECTOR:
+            bc.zero(tensor)
+        _assemble_expr(bc.f, tensor, bc.bcs, opts, assembly_rank)
 
-def collect_lgmaps(matrix, all_bcs, Vrow, Vcol, row, col):
+
+def _get_mat_type(mat_type, sub_mat_type, arguments):
+    """Provide defaults and validate matrix-type selection.
```

Comment:
Could you expand on this a little? Defaults for what? Validating what about matrix-type selection?